---
title: "Sanger-Targeted amplicon sequencing results report"
author: "Generated using [sanger_targseq](https://github.com/dhineshp565/sanger_targseq) pipeline"
date: "`r Sys.Date()`"
output: html_document
params:
  csv: ""
  
---

```{r setup, include=FALSE}
library(knitr)
library(ggplot2)
library(htmltools)
library(base64enc)
```


<br>
```{css, echo=FALSE}
.table caption {
    color: darkblue;
    font-weight: bold;
}
```

### Click dropdown to select sample {.tabset .tabset-dropdown}
```{r, results='asis',echo=FALSE}

samplelist <- read.csv(params$csv,header=FALSE,sep = ',')[-1,]
sampleid <- samplelist[,c(1)]

for (i in sampleid){
  cat("####", i, "\n")
  #use mappedreads.txt and display column 1 and 3
  
  # display abricate results
  abricate_ids <- (paste(i,"_abricate.csv",sep=""))
  abricate_raw <- read.csv(abricate_ids,header = TRUE,sep = "\t")
  abricate_final <- abricate_raw[,c(2,6,10,11,12,15)]
  colnames(abricate_final) <- c("SEQUENCE","GENE","%COVERAGE","%IDENTITY","DATABASE","REFERENCE") 
  print(knitr::kable(abricate_final,align = "llcccl",caption = "Consensus compared to reference database. Generated using [abricate](https://github.com/tseemann/abricate) and custom database")) 
  
   cat('\n\n<!-- -->\n\n')

  

  cat('\n\n<!-- -->\n\n')
  fasta_id <- (paste(i,"_consensus.fasta",sep=""))
  fasta_file <- readLines(fasta_id)
  # Display the content
  cat("```\n")
  cat(fasta_file, sep = "\n")
  cat("\n```\n")
  # Encode the file content as base64
  encoded_content <- base64encode(charToRaw(paste(fasta_file, collapse = "\n")))
  # Get file name from file path and remove extra underscores
  file_name <- gsub("_+", "_", basename(fasta_id))
  # Provide a download link for the encoded FASTA content with file name
  cat('<a href="data:text/plain;charset=utf-8;base64,', encoded_content, '" download="', file_name, '">Download FASTA file</a>')
  
  cat('\n\n<!-- -->\n\n')

  # display blast results
  blast_ids <- (paste(i,"_report_blast.csv",sep=""))
  blast_raw <- read.csv(blast_ids,header = TRUE,sep = "\t")
  blast_final <- blast_raw[,c(1,3,4,5,6,10)]
  colnames(blast_final) <- c("QUERY_ID","LENGTH","%COVERAGE","%IDENTITY","E_VAL","STITLE")
  print(knitr::kable(blast_final,align = "lccccl",caption = "Consensus compared to NCBI BLAST database. Generated using [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi)")) 
  
  cat('\n\n<!-- -->\n\n')

  # display blast results
 cat('\n\n<!-- -->\n\n')

   
  }
```




<br>


